> ### **Assignment 2 - Numpy Array Operations** 
>
> This assignment is part of the course ["Data Analysis with Python: Zero to Pandas"](http://zerotopandas.com). The objective of this assignment is to develop a solid understanding of Numpy array operations. In this assignment you will:
> 
> 1. Pick 5 interesting Numpy array functions by going through the documentation: https://numpy.org/doc/stable/reference/routines.html 
> 2. Run and modify this Jupyter notebook to illustrate their usage (some explanation and 3 examples for each function). Use your imagination to come up with interesting and unique examples.
> 3. Upload this notebook to your Jovian profile using `jovian.commit` and make a submission here: https://jovian.ml/learn/data-analysis-with-python-zero-to-pandas/assignment/assignment-2-numpy-array-operations
> 4. (Optional) Share your notebook online (on Twitter, LinkedIn, Facebook) and on the community forum thread: https://jovian.ml/forum/t/assignment-2-numpy-array-operations-share-your-work/10575 . 
> 5. (Optional) Check out the notebooks [shared by other participants](https://jovian.ml/forum/t/assignment-2-numpy-array-operations-share-your-work/10575) and give feedback & appreciation.
>
> The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.
>
> Try to give your notebook a catchy title & subtitle e.g. "All about Numpy array operations", "5 Numpy functions you didn't know you needed", "A beginner's guide to broadcasting in Numpy", "Interesting ways to create Numpy arrays", "Trigonometic functions in Numpy", "How to use Python for Linear Algebra" etc.
>
> **NOTE**: Remove this block of explanation text before submitting or sharing your notebook online - to make it more presentable.


# Numpy Array Operations


### Linear Algebra Functions

NumPy package contains numpy.linalg module that provides all the functionality required for linear algebra. I have chosen following five functions from the module: 

- `numpy.linalg.inv()`
- `numpy.linalg.det()`
- `numpy.linalg.solve()`
- `numpy.linalg.matrix_power()`
- `numpy.matmul()`

The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.

In [1]:
!pip install jovian --upgrade -q

You should consider upgrading via the 'd:\data analysis with python (jovian)\zerotopandas\scripts\python.exe -m pip install --upgrade pip' command.


In [1]:
import jovian

In [2]:
jovian.commit(project='numpy-array-operations')

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ashutoshkrris/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..


[jovian] Error: Failed to read Anaconda environment using command: "conda env export -n base --no-builds"


[jovian] Committed successfully! https://jovian.ml/ashutoshkrris/numpy-array-operations


'https://jovian.ml/ashutoshkrris/numpy-array-operations'

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [3]:
import numpy as np

In [39]:
# List of functions explained 
function1 = np.linalg.inv
function2 = np.linalg.det
function3 = np.linalg.solve
function4 = np.linalg.matrix_power
function5 = np.matmul

## Function 1 - np.linalg.inv

`np.linalg.inv()` function is used to calculate the inverse of a matrix. The inverse of a matrix is such that if it is multiplied by the original matrix, it results in identity matrix.

In [8]:
# Example 1 - working
arr1 = np.array([[1, 2], 
        [3, 4.]])

arr2 = np.linalg.inv(arr1)

arr2

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

We took a matrix arr1. We calculated the inverse of the arr1 matrix and stored it in arr2.

In [9]:
# Example 2 - working
a = np.array([[1,1,1],[0,2,5],[2,5,-1]])
a_inv = np.linalg.inv(a)
a_inv

array([[ 1.28571429, -0.28571429, -0.14285714],
       [-0.47619048,  0.14285714,  0.23809524],
       [ 0.19047619,  0.14285714, -0.0952381 ]])

In [10]:
# Example 3 - breaking (to illustrate when it breaks)
arr1 = np.array([[2,4],[1,2]])

arr2 = np.linalg.inv(arr1)

arr2

LinAlgError: Singular matrix

The inverse of a sqaure matrix does not exist when the determinant is equal to zero, i.e., the matrix is a singular matrix.

This function should be only used when the rank of a square matrix N X N is less than N or the determinant of the square matrix is not zero.

In [11]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ashutoshkrris/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..


[jovian] Error: Failed to read Anaconda environment using command: "conda env export -n base --no-builds"


[jovian] Committed successfully! https://jovian.ml/ashutoshkrris/numpy-array-operations


'https://jovian.ml/ashutoshkrris/numpy-array-operations'

## Function 2 - np.linalg.det

The `np.linalg.det()` function calculates the determinant of the input matrix.

In [14]:
# Example 1 - working
a = np.array([[1,2], [3,4]]) 
print(a)
np.linalg.det(a)

[[1 2]
 [3 4]]


-2.0000000000000004

The determinant is calculated as : `(1*4) - (2*3) = -2.0`

In [15]:
# Example 2 - working
b = np.array([[6,1,1], [4, -2, 5], [2,8,7]]) 
print(b) 
np.linalg.det(b) 

[[ 6  1  1]
 [ 4 -2  5]
 [ 2  8  7]]


-306.0

The determinant is calculated as : `6*(-2*7 - 5*8) - 1*(4*7 - 5*2) + 1*(4*8 - -2*2) = -306`

In [16]:
# Example 3 - breaking (to illustrate when it breaks)
b = np.array([[6,1,1], [4, -2, 5]]) 
print(b) 
np.linalg.det(b)

[[ 6  1  1]
 [ 4 -2  5]]


LinAlgError: Last 2 dimensions of the array must be square

Determinant of a matrix can only be calculated when the matrix is a square matrix i.e. N X N matrix.

In [17]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ashutoshkrris/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..


[jovian] Error: Failed to read Anaconda environment using command: "conda env export -n base --no-builds"


[jovian] Committed successfully! https://jovian.ml/ashutoshkrris/numpy-array-operations


'https://jovian.ml/ashutoshkrris/numpy-array-operations'

## Function 3 - np.linalg.solve

The `np.linalg.solve()` function gives the solution of linear equations in the matrix form.

In [20]:
# Example 1 - working
l1 = np.array([[2,3],[5,3]])
l2 = np.array([9,3])
sol = np.linalg.solve(l1,l2)
print(sol)

[-2.          4.33333333]


Consider the linear equations : 

    2x + 3y = 9
    5x + 3y = 3
    
After solving, we get x=-2 and y=13/3 = 4.33333

In [22]:
# Example 2 - working
a = np.array([[1,1,1],[0,2,5],[2,5,-1]])
b = np.array([6,-4,27])
print(a)
print()
print(b)
sol = np.linalg.solve(a,b)
sol

[[ 1  1  1]
 [ 0  2  5]
 [ 2  5 -1]]

[ 6 -4 27]


array([ 5.,  3., -2.])

Consider the system of linear equations :

    x + y + z = 6
    2y + 5z = -4
    2x + 5y - z = 27
    
After solving, we get x = 5, y = 3, z = -2

In [23]:
# Example 3 - breaking (to illustrate when it breaks)
a = np.array([[1,1,1],[0,2,5],[2,5,-1]])
b = np.array([6,-4,2,7])
print(a)
print()
print(b)
sol = np.linalg.solve(a,b)
sol

[[ 1  1  1]
 [ 0  2  5]
 [ 2  5 -1]]

[ 6 -4  2  7]


ValueError: solve1: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (m,m),(m)->(m) (size 4 is different from 3)

Since, there is a mismatch in the dimension of matrix a and matrix b, it could not be solved.

In [24]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ashutoshkrris/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..


[jovian] Error: Failed to read Anaconda environment using command: "conda env export -n base --no-builds"


[jovian] Committed successfully! https://jovian.ml/ashutoshkrris/numpy-array-operations


'https://jovian.ml/ashutoshkrris/numpy-array-operations'

## Function 4 - np.linalg.matrix_power

The `np.linalg.matrix_power(a,n)` raises a sqaure matrix **a** to the power **n**.

In [25]:
# Example 1 - working
a = np.array([[2,3],[5,3]])
print(a)
np.linalg.matrix_power(a,3)

[[2 3]
 [5 3]]


array([[113, 102],
       [170, 147]])

In the above example, we raise the matrix a to the power 3.

In [30]:
# Example 2 - working
a = np.zeros((4,4))
print(a)
np.linalg.matrix_power(a,0)

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

In [31]:
# Example 3 - breaking (to illustrate when it breaks)
a = np.array([[2,3],[3,4],[4,5]])
print(a)
np.linalg.matrix_power(a,3)

[[2 3]
 [3 4]
 [4 5]]


LinAlgError: Last 2 dimensions of the array must be square

This function can only be applied to square matrix.

In [32]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ashutoshkrris/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..


[jovian] Error: Failed to read Anaconda environment using command: "conda env export -n base --no-builds"


[jovian] Committed successfully! https://jovian.ml/ashutoshkrris/numpy-array-operations


'https://jovian.ml/ashutoshkrris/numpy-array-operations'

## Function 5 - np.matmul

The `np.matmul()` function returns the matrix product of two arrays. While it returns a normal product for 2-D arrays, if dimensions of either argument is >2, it is treated as a stack of matrices residing in the last two indexes and is broadcast accordingly.

In [33]:
# Example 1 - working
a = [[1,0],[0,1]] 
b = [[4,1],[2,2]] 
np.matmul(a,b)

array([[4, 1],
       [2, 2]])

For 2-D array, it is matrix multiplication

In [35]:
# Example 2 - working
a = np.arange(8).reshape(2,2,2) 
b = np.arange(4).reshape(2,2)
print("a: ",a)
print()
print("b: ",b)
np.matmul(a,b)

a:  [[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]]

b:  [[0 1]
 [2 3]]


array([[[ 2,  3],
        [ 6, 11]],

       [[10, 19],
        [14, 27]]])

For array having dimensions greater than 2

In [36]:
# Example 3 - breaking (to illustrate when it breaks)
a = np.array([1,2])
b = 3
np.matmul(a,b)

ValueError: matmul: Input operand 1 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)

Scalar multiplication raises an error.

In [37]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ashutoshkrris/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..


[jovian] Error: Failed to read Anaconda environment using command: "conda env export -n base --no-builds"


[jovian] Committed successfully! https://jovian.ml/ashutoshkrris/numpy-array-operations


'https://jovian.ml/ashutoshkrris/numpy-array-operations'

## Conclusion

In this assignment, I tried to cover some important and frequently used functions from NumPy's Linear Algebra module.

## Reference Links
Provide links to your references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* Numpy Tutorial : https://www.tutorialspoint.com/numpy/index.htm

In [40]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ashutoshkrris/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..


[jovian] Error: Failed to read Anaconda environment using command: "conda env export -n base --no-builds"


[jovian] Committed successfully! https://jovian.ml/ashutoshkrris/numpy-array-operations


'https://jovian.ml/ashutoshkrris/numpy-array-operations'